docs/proposals/ldap-group-sync.md

```diff
@@ -0,0 +1,108 @@
+
+### Use Cases
+1. I want to sync all LDAP groups I've configured in a sync job (e.g. a specific LDAP server configuration) to OpenShift `Groups`
+* I want to sync specific OpenShift `Groups` with their `externalRecordProviders`
+* I want to sync a speficic set of LDAP groups to OpenShift `Groups`
+* I want to run a destructive sync that will remove OpenShift groups that were created or updated by a previous sync if the LDAP server no longer contains their record
+* I want to map an LDAP group to an existing OpenShift `Group`
+
+## Support for LDAP Group Schemas
+The three LDAP schemas supported are
+
+1. groups as first-class LDAP entries that contain lists of members of the group
+2. users as first-class LDAP entries that contain lists of groups they are a member of
+3. schema 2 above, also including first-class groups to store metainformation
+
+### Data Flow
+* Determine list of LDAP groups to be synced
+* Find LDAP group members
+* Determine OpenShift `Identities` from LDAP group members
+* Determine OpenShift `Users` from OpenShift `Identities`
+* Determine OpenShift `Group` from LDAP group
+* Populate new OpenShift `Group`'s `Users` from resulting OpenShift `UserIdentityMappings`, update OpenShift `Group` metadata fields tied to LDAP attributes, leave other OpenShift `Group` fields unchanged
+
+##  Determining Which LDAP Groups To Sync
+An `LDAPGroupLister` determines what LDAP groups needs to be synced and outputs the result as a set of unique identifier strings (for instance, strings that encode the value of the unique filter key needed to inject into the unique filter that, when conjoined with the common filter, will retrieve those groups from the LDAP server.) The other objects that take these unique identifiers need to understand the format of this string (e.g. these objects will be tightly coupled).
+
+```go
+// LDAPGroupLister lists the specific group filters for all groups that should be synced
+type LDAPGroupLister interface {
+  ListGroups() (groupUIDs []string], err error)
+}
+```
+
+## Internalizing LDAP Responses: Building OpenShift `Users`
+### Collecting LDAP Members and Metainformation
+An `LDAPGroupDataExtractor` will be used inside of the sync job to do the first step in the sync: locate the LDAP entries that correspond to the LDAP group members of the LDAP group being synced, given the filter for that group. The `LDAPGroupDataExtractor` will also be able to retrieve the labels and annotations that should exist on the OpenShift `Group` created from this LDAP group record.
```

Comment:
An `LDAPGroupDataExtractor` gathers information for an ldap `groupUID`.  It may cache the results of LDAP queries to improve response time.